service/src/main/java/bio/terra/cbas/initialization/CloneRecoveryService.java

```diff
@@ -0,0 +1,96 @@
+package bio.terra.cbas.initialization;
+
+import bio.terra.cbas.config.CbasContextConfiguration;
+import bio.terra.cbas.dao.MethodDao;
+import bio.terra.cbas.dao.RunDao;
+import bio.terra.cbas.dao.RunSetDao;
+import bio.terra.cbas.models.Method;
+import bio.terra.cbas.models.Run;
+import bio.terra.cbas.models.RunSet;
+import bio.terra.common.db.WriteTransaction;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CloneRecoveryService {
+
+  private final Logger logger = LoggerFactory.getLogger(CloneRecoveryService.class);
+  private final RunDao runDao;
+  private final RunSetDao runSetDao;
+  private final MethodDao methodDao;
+  private final CbasContextConfiguration cbasContextConfig;
+
+  public CloneRecoveryService(
+      RunSetDao runSetDao,
+      RunDao runDao,
+      MethodDao methodDao,
+      CbasContextConfiguration cbasContextConfig) {
+    this.runSetDao = runSetDao;
+    this.runDao = runDao;
+    this.methodDao = methodDao;
+    this.cbasContextConfig = cbasContextConfig;
+  }
+
+  public void cloneRecovery() {
+    logger.info(
+        "Starting clone recovery (workspaceId: {}, workspaceCreatedDate: {})",
+        cbasContextConfig.getWorkspaceId(),
+        cbasContextConfig.getWorkspaceCreatedDate());
+
+    pruneCloneSourceWorkspaceHistory();
+  }
+
+  @WriteTransaction
+  public void pruneCloneSourceWorkspaceHistory() {
+    List<Method> methods = methodDao.getMethods().stream().filter(this::isMethodCloned).toList();
+
+    methods.forEach(this::updateMethodTemplate);
+
+    methods.forEach(this::cleanupClonedMethodHistory);
+  }
+
+  public Boolean isMethodCloned(Method m) {
+    return !m.originalWorkspaceId().equals(cbasContextConfig.getWorkspaceId());
+  }
+
+  public Boolean isRunSetCloned(RunSet r) {
+    return !r.originalWorkspaceId().equals(cbasContextConfig.getWorkspaceId());
+  }
+
+  @WriteTransaction
+  public void updateMethodTemplate(Method m) {
+    List<RunSet> methodClonedRunSets =
+        runSetDao.getRunSetsWithMethodId(m.methodId()).stream()
+            .filter(this::isRunSetCloned)
+            .toList();
+
+    if (!methodClonedRunSets.isEmpty()) {
+      RunSet newTemplateRunSet = methodClonedRunSets.get(0);
+      runSetDao.updateIsTemplate(newTemplateRunSet.runSetId(), true);
```

Comment:
I think there's a minor gotcha here. Imagine if:

- User clones a workspace with a method M and runset RS1
- User runs another instance, as RS2

Desired outcome (I think?):
- RS1 deleted
- RS2 remains in the DB

But actually, both would be kept as templates.

I don't think this is going to be a big deal _today_ because we only show the most recent run sets anyway, but if we ever give users a drop down "select from all your saved config templates" then this extra one might unexpectedly show up.

Reply:
As I understand it:
- before cloning, there are actually three runSets: RS1, RS2, and the template from the original workspace.
- we want the cloned workspace to start with the configuration of RS1, so it gets turned into a template, and the old template gets deleted
- RS2 remains in the DB as a non-template

That behavior is tested here: https://github.com/DataBiosphere/cbas/pull/261/files#diff-0a982347764cb63bb49dfee632e04a756773c56b45b8591e352748e7dd0c9399R141

We should discuss if you don't think this is the right behavior, but that was my intention here